cognify/optimizer/core/sh.py

```diff
@@ -0,0 +1,121 @@
+from typing import Callable
+import logging
+from concurrent.futures import (
+    ThreadPoolExecutor,
+    as_completed,
+)
+from concurrent.futures import Future
+
+from cognify.optimizer.evaluator import (
+    EvaluationResult,
+    GeneralEvaluatorInterface,
+)
+from cognify.optimizer.core.common_stats import CommonStats
+from cognify.optimizer.core.flow import TopDownInformation
+
+logger = logging.getLogger(__name__)
+
+class SuccessiveHalving:
+    """Policy for successive halving resource allocation
+    
+    If a layer (A) adopts this policy, it will allocate resources for the next layer (B)
+    
+    layer A will first propose `n` trials, then
+    
+    **SH routine**:
+    1. each remaining trial at layer B will have a fixed step_budget = `r_i`
+    2. after all trials are evaluated, unpromising trials will be pruned
```

Comment:
Is `prune_rate` = the number of trials that can be pruned

Reply:
nah this is prune by a factor of this rate. updated the comment